File: csharp/interval/Musarithmetic/Quality.cs
namespace Musarithmetic;

public class Quality
{
    enum kQuality{ DIMINISHED, MINOR, MAJOR, PERFECT, AUGMENTED };

    kQuality quality;

    int value { get => (int)quality; }

    public Quality(int degree, int adjustment)
    {
        quality = adjustment switch 
        {
            -2 => kQuality.DIMINISHED, 
            -1 when IsPerfectDegree(degree) => kQuality.DIMINISHED,
            -1 => kQuality.MINOR,
            0 when IsPerfectDegree(degree) => kQuality.PERFECT,
            0 => kQuality.MAJOR,
            1 => kQuality.AUGMENTED,
            _ => throw new ArgumentException($"Interval out of range")
        };
    }

    public Quality(string qualityStr)
    {
        quality = qualityStr switch
        {
            "d" => kQuality.DIMINISHED,
            "m" => kQuality.MINOR,
            "M" => kQuality.MAJOR,
            "P" => kQuality.PERFECT,
            "a" => kQuality.AUGMENTED,
            _ => throw new ArgumentException(
                    $"Unrecognized quality input {qualityStr}")
        };
    }

    public override string ToString() => 
        (new [] {"d", "m", "M", "P", "A"}).ElementAt(this.value);
    
    public int Adjustment(int degree) 
    {
        return quality switch
        {
            kQuality.DIMINISHED => IsPerfectDegree(degree) ? -1 : -2,
            kQuality.MINOR => -1,
            kQuality.MAJOR or kQuality.PERFECT => 0,
            kQuality.AUGMENTED => 1,
            _ => throw new ArgumentException(
            $"Cannot calculate chromatic value of interval quality {quality}")
        };
    }

    static bool IsPerfectDegree(int interval) => (interval % 7) is 0 or 3 or 4;

    bool IsPerfect() =>
        quality is kQuality.DIMINISHED or kQuality.PERFECT or kQuality.AUGMENTED;

    bool IsImperfect() =>
        quality is kQuality.DIMINISHED or kQuality.MINOR 
            or kQuality.MAJOR or kQuality.AUGMENTED;

    public bool MatchesDegree(int degree)
    {
        int reduced = Math.Abs(degree % 7);
        return (IsPerfect() && IsPerfectDegree(reduced))
            || (IsImperfect() && !IsPerfectDegree(reduced));
    }

}

File: java/musarithmetic/Quality.java
package com.andrewcashner.musarithmetic;

import java.util.*;

/**
 * A Quality is an inflection of an interval relative to a default base
 * value.
 * For perfect intervals the base value is PERFECT; for imperfect intervals
 * the base value is MAJOR.
 * Diminished offset values are different for imperfect and perfect
 * intervals; that code is in {@link Interval}.
 *
 * Note: we do not currently support double (or more) diminished intervals.
 */
public enum Quality implements Describable { 
    /** Diminished interval */
    DIMINISHED ("d", "diminished"),
    /** Perfect */
    PERFECT    ("P", "perfect"),
    /** Augmented */
    AUGMENTED  ("A", "augmented"),
    /** Minor */
    MINOR      ("m", "minor"),
    /** Major */
    MAJOR      ("M", "major");

    private String symbol;
    private String description;

    private Quality(String symbol, String description) {
        this.symbol = symbol;
        this.description = description;
    }

    /** Default Quality is perfect */
    public static final Quality DEFAULT = Quality.PERFECT;

    // Enable lookup by symbol
    private static final Map<String, Quality> lookup = new HashMap<>();

    static {
        Arrays.stream(Quality.values())
            .forEach(q -> lookup.put(q.symbol, q));
    }

    /**
     * Factory method to look up Quality from an input string.
     *
     * @param input User input string
     * @return New Quality instance
     * @throws IllegalArgumentException if invalid input
     */
    public static Quality parse(String input) 
            throws IllegalArgumentException {

        Quality match = lookup.get(input);
        if (input == null) {
            throw new IllegalArgumentException(String.format(
                        "Unrecognized quality input %s", input));
        }
        return match;
    }

    /**
     * String representation of the quality.
     *
     * @return String representation
     */
    public String toString() {
        return this.symbol;
    }

    public String description() {
        return this.description;
    }
} 
